iot-e2e-tests/common/src/test/java/tests/integration/com/microsoft/azure/sdk/iot/helpers/Tools.java

```diff
@@ -137,99 +154,320 @@ public static RegistryManager getRegistyManager(String iotHubConnectionString) t
+                    testModuleIdentity = new TestModuleIdentity(null, testDeviceIdentity.device, module);
+                }
+
+                ModuleClient moduleClient = new ModuleClient(DeviceConnectionString.get(iotHubConnectionString, testModuleIdentity.device, testModuleIdentity.module), protocol);
+                testModuleIdentity.setModuleClient(moduleClient);
+                return testModuleIdentity;
+            }
         }
         else
         {
-            throw new IllegalArgumentException("Test code has not been written for this path yet");
+            synchronized (testX509ModuleQueueLock)
+            {
+                TestModuleIdentity testModuleIdentity;
+                if (!needCleanTwin && testX509ModuleWithTwinQueue.size() > 0)
+                {
+                    log.debug("Acquiring test module from testX509ModuleWithTwinQueue");
+                    testModuleIdentity = testX509ModuleWithTwinQueue.remove();
+                }
+                else if (testX509ModuleQueue.size() > 0)
+                {
+                    log.debug("Acquiring test module from testX509ModuleQueue");
+                    testModuleIdentity = testX509ModuleQueue.remove();
+                }
+                else
+                {
+                    log.debug("Acquiring test module by creating new one");
+                    TestDeviceIdentity testDeviceIdentity = getTestDevice(iotHubConnectionString, protocol, authenticationType, needCleanTwin);
+                    Module module = Module.createModule(testDeviceIdentity.device.getDeviceId(), "test-module-" + UUID.randomUUID(), authenticationType);
+                    String x509Thumbprint = IntegrationTest.x509CertificateGenerator.getX509Thumbprint();
+                    module.setThumbprintFinal(x509Thumbprint, x509Thumbprint);
+                    module = addModuleWithRetry(registryManager, module);
+                    testModuleIdentity = new TestModuleIdentity(null, testDeviceIdentity.device, module);
+                }
+
+                SSLContext sslContext = SSLContextBuilder.buildSSLContext(IntegrationTest.x509CertificateGenerator.getPublicCertificate(), IntegrationTest.x509CertificateGenerator.getPrivateKey());
+                ModuleClient moduleClient = new ModuleClient(DeviceConnectionString.get(iotHubConnectionString, testModuleIdentity.device, testModuleIdentity.module), protocol, sslContext);
+                testModuleIdentity.setModuleClient(moduleClient);
+                return testModuleIdentity;
+            }
         }
     }
 
+    /**
+     * If the environment variable RECYCLE_TEST_IDENTITIES is set to true, then the identity will be recycled. Otherwise it
+     * will be removed from the IoT Hub's registry.
+     * @param testIdentity The test identity to dispose/recycle.
+     */
     public static void disposeTestIdentity(TestIdentity testIdentity, String iotHubConnectionString)
```

Comment:
Again, a bit dense, but since there are several queues of device and module identities, this function sorts the given testIdentity into the right queue